### ISSUE TITLE
firebase.json file must be present to deploy even with --config

### ISSUE THREAD

**Issue by johnelliott** [firebase/firebase-tools#2583]
### [REQUIRED] Environment info

**firebase-tools:**
8.9.2

**Platform:**
macOS node 14.9.0

## TL;DR: firebase.json seems to get read even with --config, user may want only explicit configs

firebase.json not present

firebase.elsewhere.json has this in it:
```json
{
  "hosting": {
    "public": "public",
    "ignore": [
      "firebase.json",
      "**/.*",
      "**/node_modules/**"
    ],
    "headers": [
      {
        "source": "**/*.html",
        "headers": [
          {
            "key": "Cache-Control",
            "value": "max-age=3600, stale-while-revalidate=28800"
          }
        ]
      },
      {
        "source": "**/*",
        "headers": [
          {
            "key": "X-Robots-Tag",
            "value": "noindex, nofollow, noarchive"
          }
        ]
      }
    ]
  }
}
```

### [REQUIRED] Steps to reproduce
1. run `mv  firebase.json firebase.elsewhere.json`
1. run `firebase deploy --only hosting --config firebase.elsewhere.json`

### [REQUIRED] Expected behavior

it deploys hosting

### [REQUIRED] Actual behavior
I get this error when deploying. I did verify the config file contents work when in the default path, so it doesn't seem to be the configuration file content.
```
i  hosting[userproj]: found 27 files in public
[2020-08-29T03:27:45.553Z] TypeError [ERR_INVALID_ARG_TYPE]: The "from" argument must be of type string. Received null
    at validateString (internal/validators.js:122:11)
    at Object.relative (path.js:1053:5)
    at Uploader.hashcacheName (/Users/user/.nvm/versions/node/v14.9.0/lib/node_modules/firebase-tools/lib/deploy/hosting/uploader.js:56:33)
    at new Uploader (/Users/user/.nvm/versions/node/v14.9.0/lib/node_modules/firebase-tools/lib/deploy/hosting/uploader.js:49:60)
    at _runDeploys (/Users/user/.nvm/versions/node/v14.9.0/lib/node_modules/firebase-tools/lib/deploy/hosting/deploy.js:31:26)
    at module.exports (/Users/user/.nvm/versions/node/v14.9.0/lib/node_modules/firebase-tools/lib/deploy/hosting/deploy.js:64:12)
    at _chain (/Users/user/.nvm/versions/node/v14.9.0/lib/node_modules/firebase-tools/lib/deploy/index.js:22:40)
    at /Users/user/.nvm/versions/node/v14.9.0/lib/node_modules/firebase-tools/lib/deploy/index.js:66:16
    at processTicksAndRejections (internal/process/task_queues.js:93:5)

Error: An unexpected error has occurred.
```

My current work-around is to just `touch firebase.json` to have an empty file, which gives me the error I want error when --config is not set.

I may have time to debug this at some point since I use firebase a lot, but hopefully if you work on this full time it'll be obvious what's going on.

**Comment by samtstern** [comment]
This looks to be related to https://github.com/firebase/firebase-tools/issues/2569
